pyproject.toml

```diff
@@ -167,6 +167,27 @@ extend_skip_glob = ["asdf/extern/*", ".eggs/*", ".tox/*"]
 
 [tool.ruff]
 line-length = 120
+select = ["ALL"]
+extend-ignore = [
+    # Ignored check groups
+    "C90", # mccabe complexity
```

Comment:
This check is interesting. It runs the checks in https://github.com/pycqa/mccabe, which is running the [Cyclomatic complexity](https://en.wikipedia.org/wiki/Cyclomatic_complexity) metric on our code. It then flags areas which are too complex in the metric.

It might be a good idea for us to explore simplifying/refactoring the regions of our code this flags.

Reply:
Errors found by this:
```shell
asdf/_node_info.py:240:9: C901 `from_root_node` is too complex (11)
asdf/asdf.py:283:9: C901 `_check_extensions` is too complex (12)
asdf/asdf.py:403:9: C901 `_update_extension_history` is too complex (11)
asdf/asdf.py:1456:9: C901 `add_history_entry` is too complex (12)
asdf/block.py:393:9: C901 `read_block_index` is too complex (19)
asdf/block.py:597:9: C901 `get_block` is too complex (14)
asdf/block.py:972:9: C901 `read` is too complex (19)
asdf/block.py:1277:5: C901 `calculate_updated_layout` is too complex (13)
asdf/commands/edit.py:177:5: C901 `edit` is too complex (20)
asdf/commands/extension.py:54:5: C901 `_print_extension_details` is too complex (13)
asdf/extension/_extension.py:135:9: C901 `__init__` is too complex (13)
asdf/extension/_manager.py:20:9: C901 `__init__` is too complex (12)
asdf/generic_io.py:121:9: C901 `read` is too complex (12)
asdf/generic_io.py:952:5: C901 `get_file` is too complex (21)
asdf/search.py:114:9: C901 `search` is too complex (14)
asdf/search.py:167:13: C901 `_filter` is too complex (12)
asdf/tags/core/ndarray.py:39:5: C901 `asdf_datatype_to_numpy_dtype` is too complex (11)
asdf/tags/core/ndarray.py:132:5: C901 `inline_data_asarray` is too complex (12)
asdf/tags/core/ndarray.py:401:9: C901 `to_tree` is too complex (12)
asdf/tags/core/ndarray.py:647:5: C901 `validate_datatype` is too complex (15)
asdf/tests/helpers.py:73:5: C901 `assert_tree_match` is too complex (15)
asdf/tests/helpers.py:106:9: C901 `recurse` is too complex (12)
asdf/tests/helpers.py:432:5: C901 `_assert_extension_type_correctness` is too complex (11)
asdf/treeutil.py:221:5: C901 `walk_and_modify` is too complex (30)
asdf/types.py:85:9: C901 `__new__` is too complex (17)
asdf/yamlutil.py:211:5: C901 `custom_tree_to_tagged_tree` is too complex (11)
asdf/yamlutil.py:272:5: C901 `tagged_tree_to_custom_tree` is too complex (11)
```